roles/jd-server/src/lib/job_declarator/mod.rs

```diff
@@ -450,25 +453,62 @@ impl JobDeclarator {
             if let Ok((receiver, sender, _, _)) =
                 Connection::new(stream, HandshakeRole::Responder(responder)).await
             {
-                let setup_message_from_proxy_jd = receiver.recv().await.unwrap();
-                info!(
-                    "Setup connection message from proxy: {:?}",
-                    setup_message_from_proxy_jd
-                );
-
-                let setup_connection_success_to_proxy = SetupConnectionSuccess {
-                    used_version: 2,
-                    // Setup flags for async_mining_allowed
-                    flags: 0b_0000_0000_0000_0000_0000_0000_0000_0001,
+                let address = addr.unwrap();
+                let mut incoming = match receiver.recv().await {
+                    Ok(EitherFrame::Sv2(s)) => {
+                        debug!("Got sv2 message: {:?}", s);
+                        s
+                    }
+                    Ok(EitherFrame::HandShake(s)) => {
+                        error!(
+                            "Got unexpected handshake message from upstream: {:?} at {}",
+                            s, address
+                        );
+                        continue;
+                    }
+                    Err(e) => {
+                        error!("Error receiving message: {:?}", e);
+                        continue;
+                    }
                 };
-                let sv2_frame: StdFrame =
-                    JdsMessages::Common(setup_connection_success_to_proxy.into())
+
+                let payload = incoming.payload();
+                let flag = payload[5] as u32;
```

Comment:
shouldn't we do some kind of verification to make sure this is indeed a `SetupConnection` that JDS just received?

what if the client sent some nonsense and we're blindly parsing this payload as if it was a `SetupConnection` flag?

Reply:
also, what is the rationale behind `payload[5]`?

Reply:
I completely overlooked this possibility. It makes sense to have this approach rather than parsing based on my understanding. The reason for payload[5] was that the flag is a 4-byte value encoded in the payload in little-endian format, so the 5th byte must contain the flag value. Since the current range for the flag value is up to 3 bits max, it fits easily within 8 bits or the 1st byte. However, we still have the possibility of increasing the flag bit range in the future.

For any other type of message except the setupConnection, we should not respond. Instead, we should continue listening to our other processes, as these messages might be some sort of attack to slow down our JDS server. In such cases, we should avoid sending any response to the connection request.

But ya thanks for pointing this out. @plebhash

Reply:
> The reason for payload[5] was that the flag is a 4-byte value encoded in the payload in little-endian format, so the 5th byte must contain the flag value.

we have APIs for parsing messages from frame payloads, so we should use them for safety and readability

> For any other type of message except the setupConnection, we should not respond. Instead, we should continue listening to our other processes, as these messages might be some sort of attack to slow down our JDS server. In such cases, we should avoid sending any response to the connection request.

the attack vector is a good point that I also overlooked in my previous message.

given that the context of this code is a `while let Ok((stream, _)) = listner.accept().await { ... }`, I wonder if it doesn't make sense to just drop the connection for every case where the first incoming message is not a `SetupConnection`?

cc @Fi3